File: frontend/src/renderer/src/stores/repo-source.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import type { RepoSource, RepoSourceType, RepoSubMarket, MarketplaceItem } from '../types/marketplace'
import { useApi } from '../composables/useApi'

const REPO_SOURCES_KEY = 'luominest-repo-sources-active'

export interface SyncedItemsResult {
  items: MarketplaceItem[]
  total: number
  sourceId: string
  subMarketId?: string
  syncedAt?: string
  fromCache?: boolean
}

export const useRepoSourceStore = defineStore('repoSource', () => {
  const sources = ref<RepoSource[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const activeSourceId = ref<string>(loadActiveSourceId())

  // 同步后的远程条目缓存（按 sourceId 索引）
  const syncedItems = ref<Record<string, MarketplaceItem[]>>({})
  const syncedItemsLoading = ref<Record<string, boolean>>({})

  function loadActiveSourceId(): string {
    try {
      const stored = localStorage.getItem(REPO_SOURCES_KEY)
      if (stored && sources.value.some(s => s.id === stored)) {
        return stored
      }
      return sources.value[0]?.id || ''
    } catch {
      return sources.value[0]?.id || ''
    }
  }

  const saveActiveSourceId = (id: string) => {
    localStorage.setItem(REPO_SOURCES_KEY, id)
  }

  const activeSource = computed(() => {
    return sources.value.find(s => s.id === activeSourceId.value) || sources.value[0] || null
  })

  const enabledSources = computed(() => {
    return sources.value.filter(s => s.enabled)
  })

  const githubSources = computed(() => {
    return sources.value.filter(s => s.type === 'github')
  })

  const cloudSources = computed(() => {
    return sources.value.filter(s => s.type === 'cloud')
  })

  const cdnSources = computed(() => {
    return sources.value.filter(s => s.type === 'cdn')
  })

  const customSources = computed(() => {
    return sources.value.filter(s => s.type === 'custom')
  })

  const groupedSources = computed(() => {
    const groups: { type: RepoSourceType; label: string; icon: string; items: RepoSource[] }[] = [
      { type: 'github', label: 'GitHub', icon: 'Github', items: githubSources.value },
      { type: 'cloud', label: '云端', icon: 'Cloud', items: cloudSources.value },
      { type: 'cdn', label: 'CDN', icon: 'Globe', items: cdnSources.value },
      { type: 'custom', label: '自定义', icon: 'Plus', items: customSources.value },
    ]
    return groups.filter(g => g.items.length > 0)
  })

  // 当前活跃来源的已同步条目
  const activeSourceItems = computed<MarketplaceItem[]>(() => {
    return syncedItems.value[activeSourceId.value] || []
  })

  const fetchSources = async () => {
    loading.value = true
    error.value = null
    try {
      const api = useApi()
      const data = await api.apiGet<RepoSource[]>('/repo-sources')
      sources.value = data
    } catch (e: any) {
      error.value = e.message
      sources.value = getDefaultSources()
    } finally {
      loading.value = false
    }
  }

  const setActiveSource = (sourceId: string) => {
    activeSourceId.value = sourceId
    saveActiveSourceId(sourceId)
    // 切换来源时自动加载缓存条目
    fetchSourceItems(sourceId)
  }

  const toggleSource = async (sourceId: string) => {
    try {
      const api = useApi()
      const updated = await api.apiPost<RepoSource>(`/repo-sources/${sourceId}/toggle`)
      const idx = sources.value.findIndex(s => s.id === sourceId)
      if (idx >= 0) {
        sources.value[idx] = updated
      }
    } catch (e: any) {
      error.value = e?.message || '操作失败'
    }
  }

  const unlinkSubMarket = async (sourceId: string, subMarketId: string) => {
    try {
      const api = useApi()
      const updated = await api.apiPatch<RepoSource>(`/repo-sources/${sourceId}/sub-markets/${subMarketId}/unlink`)
      const idx = sources.value.findIndex(s => s.id === sourceId)
      if (idx >= 0) {
        sources.value[idx] = updated
      }
    } catch (e: any) {
      error.value = e?.message || '操作失败'
    }
  }

  const linkSubMarket = async (sourceId: string, subMarketId: string) => {
    try {
      const api = useApi()
      const updated = await api.apiPatch<RepoSource>(`/repo-sources/${sourceId}/sub-markets/${subMarketId}/link`)
      const idx = sources.value.findIndex(s => s.id === sourceId)
      if (idx >= 0) {
        sources.value[idx] = updated
      }
    } catch (e: any) {
      error.value = e?.message || '操作失败'
    }
  }

  const syncSource = async (sourceId: string, force: boolean = false) => {
    try {
      const api = useApi()
      const idx = sources.value.findIndex(s => s.id === sourceId)
      if (idx >= 0) {
        sources.value[idx] = { ...sources.value[idx], status: 'loading' }
      }
      syncedItemsLoading.value = { ...syncedItemsLoading.value, [sourceId]: true }

      const updated = await api.apiPost<RepoSource>(`/repo-sources/${sourceId}/sync${force ? '?force=true' : ''}`)
      if (idx >= 0) {
        sources.value[idx] = updated
      }

      // 同步完成后获取条目
      await fetchSourceItems(sourceId)
    } catch (e: any) {
      error.value = e.message
      const idx = sources.value.findIndex(s => s.id === sourceId)
      if (idx >= 0) {
        sources.value[idx] = { ...sources.value[idx], status: 'error', errorMessage: e.message }
      }
    } finally {
      syncedItemsLoading.value = { ...syncedItemsLoading.value, [sourceId]: false }
    }
  }

  const syncSubMarket = async (sourceId: string, subMarketId: string, force: boolean = false) => {
    try {
      const api = useApi()
      const result = await api.apiPost<SyncedItemsResult>(
        `/repo-sources/${sourceId}/sub-markets/${subMarketId}/sync${force ? '?force=true' : ''}`
      )
      // 更新该来源的条目缓存
      await fetchSourceItems(sourceId)
      return result
    } catch (e: any) {
      error.value = e.message
      return null
    }
  }

  /**
   * 获取仓库来源下的已缓存市场条目（不触发网络请求，仅读取后端缓存）
   */
  const fetchSourceItems = async (sourceId: string) => {
    try {
      const api = useApi()
      const result = await api.apiGet<SyncedItemsResult>(`/repo-sources/${sourceId}/items`)
      syncedItems.value = { ...syncedItems.value, [sourceId]: result.items || [] }
    } catch (e: any) {
      // 静默失败，不影响用户体验
      console.warn('[RepoSource] Failed to fetch source items:', e.message)
    }
  }

  /**
   * 获取仓库来源下指定类型的市场条目
   */
  const fetchSourceItemsByType = async (sourceId: string, type: string) => {
    try {
      const api = useApi()
      const result = await api.apiGet<SyncedItemsResult>(`/repo-sources/${sourceId}/items?type=${type}`)
      return result.items || []
    } catch (e: any) {
      console.warn('[RepoSource] Failed to fetch source items by type:', e.message)
      return []
    }
  }

  /**
   * 清除仓库来源的缓存
   */
  const clearSourceCache = async (sourceId: string) => {
    try {
      const api = useApi()
      await api.apiDelete(`/repo-sources/${sourceId}/cache`)
      // 清除本地缓存
      const newSynced = { ...syncedItems.value }
      delete newSynced[sourceId]
      syncedItems.value = newSynced
    } catch (e: any) {
      error.value = e.message
    }
  }

  const addCustomSource = async (data: { name: string; url: string; description?: string }) => {
    try {
      const api = useApi()
      const created = await api.apiPost<RepoSource>('/repo-sources', {
        type: 'custom',
        name: data.name,
        url: data.url,
        description: data.description || '',
        enabled: true,
        sub_markets: [],
      })
      sources.value.push(created)
      return created
    } catch (e: any) {
      error.value = e.message
      return null
    }
  }

  const deleteSource = async (sourceId: string) => {
    try {
      const api = useApi()
      await api.apiDelete(`/repo-sources/${sourceId}`)
      sources.value = sources.value.filter(s => s.id !== sourceId)
      // 清除本地条目缓存
      const newSynced = { ...syncedItems.value }
      delete newSynced[sourceId]
      syncedItems.value = newSynced
      if (activeSourceId.value === sourceId) {
        activeSourceId.value = sources.value[0]?.id || ''
        saveActiveSourceId(activeSourceId.value)
      }
    } catch (e: any) {
      error.value = e.message
    }
  }

  function getDefaultSources(): RepoSource[] {
    return [
      {
        id: 'github-official',
        type: 'github',
        name: 'GitHub 官方仓库',
        url: 'https://github.com/LuminousCX',
        description: 'LuminousCX 官方 GitHub 仓库，包含技能、智能体和插件',
        enabled: true,
        subMarkets: [
          { id: 'gh-skills', name: 'Skills 市场', type: 'skill', url: 'https://github.com/LuminousCX/skills', description: 'LuminousCX 官方技能仓库', linked: true },
          { id: 'gh-agents', name: 'Agents 市场', type: 'agent', url: 'https://github.com/LuminousCX/agents', description: 'LuminousCX 官方智能体仓库', linked: true },
          { id: 'gh-plugins', name: 'Plugins 市场', type: 'plugin', url: 'https://github.com/LuminousCX/plugins', description: 'LuminousCX 官方插件仓库', linked: true },
        ],
        status: 'idle',
      },
      {
        id: 'cloud-official',
        type: 'cloud',
        name: '云端仓库',
        description: 'LuminousCX 云端托管仓库',
        enabled: false,
        subMarkets: [],
        status: 'idle',
      },
      {
        id: 'cdn-official',
        type: 'cdn',
        name: 'CDN 仓库',
        description: 'CDN 加速分发仓库',
        enabled: false,
        subMarkets: [],
        status: 'idle',
      },
    ]
  }

  return {
    sources,
    loading,
    error,
    activeSourceId,
    activeSource,
    enabledSources,
    githubSources,
    cloudSources,
    cdnSources,
    customSources,
    groupedSources,
    syncedItems,
    syncedItemsLoading,
    activeSourceItems,
    fetchSources,
    setActiveSource,
    toggleSource,
    unlinkSubMarket,
    linkSubMarket,
    syncSource,
    syncSubMarket,
    fetchSourceItems,
    fetchSourceItemsByType,
    clearSourceCache,
    addCustomSource,
    deleteSource,
  }
})

File: backend/app/api/v1/endpoints/repo_source.py
import uuid
from datetime import datetime, timezone
from typing import Optional
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field, ConfigDict
from loguru import logger

from app.infrastructure.database.json_store import repo_sources_store

router = APIRouter(prefix="/repo-sources", tags=["repo-sources"])


class SubMarketCreate(BaseModel):
    name: str
    type: str = Field(..., pattern=r"^(plugin|skill|agent)$")
    url: str
    description: str = ""


class SubMarketUpdate(BaseModel):
    name: Optional[str] = None
    url: Optional[str] = None
    description: Optional[str] = None
    linked: Optional[bool] = None


class SubMarketResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True, by_alias=True)

    id: str
    name: str
    type: str
    url: str
    description: str = ""
    linked: bool = True


class RepoSourceCreate(BaseModel):
    type: str = Field(..., pattern=r"^(github|cloud|cdn|custom)$")
    name: str
    url: str = ""
    description: str = ""
    enabled: bool = True
    sub_markets: list[SubMarketCreate] = Field(default_factory=list)


class RepoSourceUpdate(BaseModel):
    name: Optional[str] = None
    url: Optional[str] = None
    description: Optional[str] = None
    enabled: Optional[bool] = None
    sub_markets: Optional[list[SubMarketCreate]] = None


class RepoSourceResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True, by_alias=True)

    id: str
    type: str
    name: str
    url: str = ""
    description: str = ""
    enabled: bool = True
    sub_markets: list[SubMarketResponse] = Field(alias="subMarkets", default_factory=list)
    last_synced_at: str = Field(alias="lastSyncedAt", default="")
    status: str = "idle"
    error_message: str = Field(alias="errorMessage", default="")


DEFAULT_GITHUB_SUB_MARKETS = [
    {"id": "gh-skills", "name": "Skills 市场", "type": "skill", "url": "https://github.com/LuminousCX/skills", "description": "LuminousCX 官方技能仓库", "linked": True},
    {"id": "gh-agents", "name": "Agents 市场", "type": "agent", "url": "https://github.com/LuminousCX/agents", "description": "LuminousCX 官方智能体仓库", "linked": True},
    {"id": "gh-plugins", "name": "Plugins 市场", "type": "plugin", "url": "https://github.com/LuminousCX/plugins", "description": "LuminousCX 官方插件仓库", "linked": True},
]

DEFAULT_REPO_SOURCES = [
    {
        "id": "github-official",
        "type": "github",
        "name": "GitHub 官方仓库",
        "url": "https://github.com/LuminousCX",
        "description": "LuminousCX 官方 GitHub 仓库，包含技能、智能体和插件",
        "enabled": True,
        "sub_markets": DEFAULT_GITHUB_SUB_MARKETS,
        "last_synced_at": "",
        "status": "idle",
        "error_message": "",
    },
    {
        "id": "cloud-official",
        "type": "cloud",
        "name": "云端仓库",
        "url": "",
        "description": "LuminousCX 云端托管仓库",
        "enabled": False,
        "sub_markets": [],
        "last_synced_at": "",
        "status": "idle",
        "error_message": "",
    },
    {
        "id": "cdn-official",
        "type": "cdn",
        "name": "CDN 仓库",
        "url": "",
        "description": "CDN 加速分发仓库",
        "enabled": False,
        "sub_markets": [],
        "last_synced_at": "",
        "status": "idle",
        "error_message": "",
    },
]


def _ensure_defaults():
    existing = repo_sources_store.all()
    if not existing:
        for source in DEFAULT_REPO_SOURCES:
            repo_sources_store.set(source["id"], source)
        logger.info("[RepoSources] Initialized default repo sources")


_ensure_defaults()


def _to_response(source: dict) -> RepoSourceResponse:
    sub_markets = [
        SubMarketResponse(
            id=sm.get("id", ""),
            name=sm.get("name", ""),
            type=sm.get("type", "plugin"),
            url=sm.get("url", ""),
            description=sm.get("description", ""),
            linked=sm.get("linked", True),
        )
        for sm in source.get("sub_markets", [])
    ]
    return RepoSourceResponse(
        id=source.get("id", ""),
        type=source.get("type", "github"),
        name=source.get("name", ""),
        url=source.get("url", ""),
        description=source.get("description", ""),
        enabled=source.get("enabled", True),
        sub_markets=sub_markets,
        last_synced_at=source.get("last_synced_at", ""),
        status=source.get("status", "idle"),
        error_message=source.get("error_message", ""),
    )


@router.get("", response_model=list[RepoSourceResponse])
async def list_repo_sources():
    logger.info("[API] GET /repo-sources - Listing all repo sources")
    sources = repo_sources_store.all()
    return [_to_response(s) for s in sources]


@router.post("", response_model=RepoSourceResponse)
async def create_repo_source(request: RepoSourceCreate):
    logger.info(f"[API] POST /repo-sources - Creating repo source: name={request.name}")
    source_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc).isoformat()
    sub_markets = []
    for sm in request.sub_markets:
        sub_markets.append({
            "id": str(uuid.uuid4())[:8],
            "name": sm.name,
            "type": sm.type,
            "url": sm.url,
            "description": sm.description,
            "linked": True,
        })
    source = {
        "id": source_id,
        "type": request.type,
        "name": request.name,
        "url": request.url,
        "description": request.description,
        "enabled": request.enabled,
        "sub_markets": sub_markets,
        "last_synced_at": "",
        "status": "idle",
        "error_message": "",
        "created_at": now,
        "updated_at": now,
    }
    repo_sources_store.set(source_id, source)
    logger.success(f"[API] POST /repo-sources - Created: id={source_id}")
    return _to_response(source)


@router.get("/{source_id}", response_model=RepoSourceResponse)
async def get_repo_source(source_id: str):
    logger.info(f"[API] GET /repo-sources/{source_id}")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")
    return _to_response(source)


@router.patch("/{source_id}", response_model=RepoSourceResponse)
async def update_repo_source(source_id: str, request: RepoSourceUpdate):
    logger.info(f"[API] PATCH /repo-sources/{source_id}")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    update_data = request.model_dump(exclude_unset=True)
    if "sub_markets" in update_data and update_data["sub_markets"] is not None:
        existing_sms = source.get("sub_markets", [])
        update_data["sub_markets"] = [
            {
                "id": existing_sms[i]["id"] if i < len(existing_sms) else str(uuid.uuid4())[:8],
                "name": sm["name"],
                "type": sm["type"],
                "url": sm["url"],
                "description": sm.get("description", ""),
                "linked": True,
            }
            for i, sm in enumerate(update_data["sub_markets"])
        ]
    source.update(update_data)
    source["updated_at"] = datetime.now(timezone.utc).isoformat()
    repo_sources_store.set(source_id, source)
    logger.success(f"[API] PATCH /repo-sources/{source_id} - Updated")
    return _to_response(source)


@router.delete("/{source_id}")
async def delete_repo_source(source_id: str):
    logger.info(f"[API] DELETE /repo-sources/{source_id}")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")
    repo_sources_store.delete(source_id)
    logger.success(f"[API] DELETE /repo-sources/{source_id} - Deleted")
    return {"error": None, "data": {"deleted": True}}


@router.post("/{source_id}/toggle", response_model=RepoSourceResponse)
async def toggle_repo_source(source_id: str):
    logger.info(f"[API] POST /repo-sources/{source_id}/toggle")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")
    source["enabled"] = not source.get("enabled", True)
    source["updated_at"] = datetime.now(timezone.utc).isoformat()
    repo_sources_store.set(source_id, source)
    logger.success(f"[API] POST /repo-sources/{source_id}/toggle - enabled={source['enabled']}")
    return _to_response(source)


@router.patch("/{source_id}/sub-markets/{sub_market_id}/unlink")
async def unlink_sub_market(source_id: str, sub_market_id: str):
    logger.info(f"[API] PATCH /repo-sources/{source_id}/sub-markets/{sub_market_id}/unlink")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    sub_markets = source.get("sub_markets", [])
    found = False
    for sm in sub_markets:
        if sm.get("id") == sub_market_id:
            sm["linked"] = False
            found = True
            break
    if not found:
        raise HTTPException(status_code=404, detail=f"Sub-market {sub_market_id} not found")

    source["sub_markets"] = sub_markets
    source["updated_at"] = datetime.now(timezone.utc).isoformat()
    repo_sources_store.set(source_id, source)
    logger.success(f"[API] Unlinked sub-market {sub_market_id}")
    return _to_response(source)


@router.patch("/{source_id}/sub-markets/{sub_market_id}/link")
async def link_sub_market(source_id: str, sub_market_id: str):
    logger.info(f"[API] PATCH /repo-sources/{source_id}/sub-markets/{sub_market_id}/link")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    sub_markets = source.get("sub_markets", [])
    found = False
    for sm in sub_markets:
        if sm.get("id") == sub_market_id:
            sm["linked"] = True
            found = True
            break
    if not found:
        raise HTTPException(status_code=404, detail=f"Sub-market {sub_market_id} not found")

    source["sub_markets"] = sub_markets
    source["updated_at"] = datetime.now(timezone.utc).isoformat()
    repo_sources_store.set(source_id, source)
    logger.success(f"[API] Linked sub-market {sub_market_id}")
    return _to_response(source)


@router.post("/{source_id}/sync", response_model=RepoSourceResponse)
async def sync_repo_source(source_id: str, force: bool = False):
    logger.info(f"[API] POST /repo-sources/{source_id}/sync")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    source["status"] = "loading"
    source["error_message"] = ""
    repo_sources_store.set(source_id, source)

    try:
        from app.infrastructure.sync.github_sync import sync_source
        results = await sync_source(source_id, source, force=force)

        # 检查是否有错误
        errors = [r.error for r in results if r.error]
        if errors and len(errors) == len(results):
            # 所有子市场都失败
            source["status"] = "error"
            source["error_message"] = "; ".join(errors)
        elif errors:
            # 部分失败
            source["status"] = "loaded"
            source["error_message"] = f"部分子市场同步失败: {'; '.join(errors)}"
        else:
            source["status"] = "loaded"
            source["error_message"] = ""

        source["last_synced_at"] = datetime.now(timezone.utc).isoformat()
        source["updated_at"] = datetime.now(timezone.utc).isoformat()
        repo_sources_store.set(source_id, source)
        logger.success(f"[API] POST /repo-sources/{source_id}/sync - Success ({len(results)} sub-markets)")
    except Exception as e:
        source["status"] = "error"
        source["error_message"] = str(e)
        repo_sources_store.set(source_id, source)
        logger.error(f"[API] POST /repo-sources/{source_id}/sync - Error: {e}")

    return _to_response(source)


@router.post("/{source_id}/sub-markets/{sub_market_id}/sync")
async def sync_sub_market(source_id: str, sub_market_id: str, force: bool = False):
    """同步单个子市场"""
    logger.info(f"[API] POST /repo-sources/{source_id}/sub-markets/{sub_market_id}/sync")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    sub_market = None
    for sm in source.get("sub_markets", []):
        if sm.get("id") == sub_market_id:
            sub_market = sm
            break
    if not sub_market:
        raise HTTPException(status_code=404, detail=f"Sub-market {sub_market_id} not found")

    try:
        from app.infrastructure.sync.github_sync import sync_sub_market as do_sync
        result = await do_sync(source_id, sub_market_id, sub_market.get("url", ""), force=force)
        return result.to_dict()
    except Exception as e:
        logger.error(f"[API] Sync sub-market failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{source_id}/items")
async def get_source_items(source_id: str, type: Optional[str] = None):
    """获取仓库来源下的所有已缓存市场条目"""
    logger.info(f"[API] GET /repo-sources/{source_id}/items")
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    from app.infrastructure.sync.github_sync import get_all_cached_items
    items = get_all_cached_items(source_id, source)

    # 按类型过滤
    if type and type in ("plugin", "skill", "agent"):
        items = [i for i in items if i.get("type") == type]

    return {"items": items, "total": len(items), "sourceId": source_id}


@router.get("/{source_id}/sub-markets/{sub_market_id}/items")
async def get_sub_market_items(source_id: str, sub_market_id: str):
    """获取子市场下的已缓存市场条目"""
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    from app.infrastructure.sync.github_sync import get_cached_items
    cached = get_cached_items(source_id, sub_market_id)
    if cached:
        return {"items": cached.items, "total": cached.total, "sourceId": source_id, "subMarketId": sub_market_id, "syncedAt": cached.synced_at, "fromCache": cached.from_cache}
    return {"items": [], "total": 0, "sourceId": source_id, "subMarketId": sub_market_id}


@router.delete("/{source_id}/cache")
async def clear_source_cache(source_id: str, sub_market_id: Optional[str] = None):
    """清除仓库来源的缓存"""
    source = repo_sources_store.get(source_id)
    if not source:
        raise HTTPException(status_code=404, detail=f"Repo source {source_id} not found")

    from app.infrastructure.sync.github_sync import clear_cache
    clear_cache(source_id=source_id, sub_market_id=sub_market_id)
    return {"error": None, "data": {"cleared": True}}
